compiler/rustc_middle/src/mir/consts.rs

```diff
@@ -498,22 +498,19 @@ impl<'tcx> Const<'tcx> {
         }
     }
 
-    /// Return true if any evaluation of this constant returns the same value.
+    /// Return true if any evaluation of this constant always returns the same value,
+    /// taking into account even pointer identity tests.
     pub fn is_deterministic(&self) -> bool {
         // Some constants may contain pointers. We need to preserve the provenance of these
         // pointers, but not all constants guarantee this:
         // - valtrees purposefully do not;
         // - ConstValue::Slice does not either.
         match self {
             Const::Ty(c) => match c.kind() {
-                ty::ConstKind::Value(valtree) => match valtree {
-                    // This is just an integer, keep it.
-                    ty::ValTree::Leaf(_) => true,
-                    // This branch may be a reference. Valtree references correspond to a
-                    // different allocation each time they are evaluated.
-                    ty::ValTree::Branch(_) => false,
-                },
                 ty::ConstKind::Param(..) => true,
+                // A valtree may be a reference. Valtree references correspond to a
+                // different allocation each time they are evaluated.
+                ty::ConstKind::Value(_) => false,
```

Comment:
Do you have the type available here? Valtrees of types that satisfy `is_primitive` should be good.